Fix part of #7305 separate LintChecksManager

cuichenli opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Separate the LintChecksManager to JsTsLintChecksManager and OtherLintChecksManager. Make LintChecksManager as base class of the other two. It has methods to do some common checkes against (almost) all files, JsTsLintChecksManger would run the common checks and some specific lints against js and ts files. OtherLinkChecksManager do the checks against other files. 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

cuichenli committed:
Separate LintChecksManager

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into 7305-separate-lintchecksmanager

cuichenli committed:
pylint

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1140,7 +1140,7 @@ def _lint_py_files(
 
 
 class LintChecksManager(object):
```
Since we don't want this to be instantiated directly, we want to have this be an abstract class.

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1140,7 +1140,7 @@ def _lint_py_files(
 
 class LintChecksManager(object):
-    """Manages all the linting functions.
+    """Manages all the common linting functions.
```
Also mention here that this class isn't intended to be used directly.

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1165,76 +1165,6 @@ def __init__(self, all_filepaths, verbose_mode_enabled=False):
-            dir_path, os.path.basename(filepath).replace('.ts', '.js'))
-        return compiled_js_filepath
 
     def _lint_all_files(self):
```
This method should be broken up and implemented in the child classes, so JS/TS lint in one class, and the everything else in the other class.

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1612,262 +1362,1049 @@ def _check_mandatory_patterns(self):
+        self.all_filepaths = js_filepaths + ts_filepaths
+        self.js_filepaths = js_filepaths
+        self.ts_filepaths = ts_filepaths
+        self.verbose_mode_enabled = verbose_mode_enabled
```
This shouldn't be needed since `verbose_mode_enabled` is already defined in the parent.

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1612,262 +1362,1049 @@ def _check_mandatory_patterns(self):
+
+
+class OtherLintChecksManager(LintChecksManager):
+    """Manages all the Other linting functions.
```
Note here that we aren't checking JS/TS files. If possible, be explicit in what we are linting.

cuichenli replied:
Done

kevinlee12 reviewed (changes requested):
Hi @cuichenli, thanks for the PR, I've left a couple comments, ptal!

@ankita240796, can you also do a pass at this to make sure I haven't missed anything?

ankita240796 commented:
Hi @cuichenli, thanks for the PR! 

@kevinlee12 I have gone through the PR and I think you have covered the issues in your comments. There is some common code in js and ts lint checks but I am working on removing code repetition and this PR is focussed on separation of tests only. So, I think it looks good overall. Thanks!

cuichenli committed:
Fix issues

- Make LintChecksManager abstract class
- Separate _lint_all_files
- Remove redundtant verbose_mode_enabled
- Add comments

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1165,19 +1168,422 @@ def __init__(self, all_filepaths, verbose_mode_enabled=False):
+        self.all_filepaths = js_filepaths + ts_filepaths
+        self.js_filepaths = js_filepaths
+        self.ts_filepaths = ts_filepaths
+        LintChecksManager.__init__(
```
I believe calling it the parent class like this might cause some issues, use [super](https://docs.python.org/2/library/functions.html#super) instead.

cuichenli replied:
Updated in the new commit

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1165,19 +1168,422 @@ def __init__(self, all_filepaths, verbose_mode_enabled=False):
+    def _lint_all_files(self):
+        """This function is used to check if node-eslint dependencies are
+        installed and pass ESLint binary path and lint all the files(JS, Python,
+        HTML, CSS) with their respective third party linters.
```
Update the comment, since it doesn't lint Python, HTML, or CSS files any more :smile:

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1165,19 +1168,422 @@ def __init__(self, all_filepaths, verbose_mode_enabled=False):
+        HTML, CSS) with their respective third party linters.
+        """
+
+        print 'Starting linter...'
```
Note there that the linter is for JS and TS files.

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1806,23 +2247,152 @@ def _check_html_directive_name(self):
+        self.css_filepaths = css_filepaths
+        self.all_filepaths = (
+            py_filepaths + html_filepaths + other_filepaths + css_filepaths)
+        LintChecksManager.__init__(
```
Ditto

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1806,23 +2247,152 @@ def _check_html_directive_name(self):
+
+    def _lint_all_files(self):
+        """This function is used to check if node-eslint dependencies are
+        installed and pass ESLint binary path and lint all the files(JS, Python,
```
Not JS files right?

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1806,23 +2247,152 @@ def _check_html_directive_name(self):
+        HTML, CSS) with their respective third party linters.
+        """
+
+        print 'Starting linter...'
```
Note here that it is for CSS, HTML, and Python files.

cuichenli replied:
Done

kevinlee12 reviewed (changes requested):
thanks @cuichenli, I have a couple more comments, ptal!

cuichenli committed:
Update comments and std out

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into 7305-separate-lintchecksmanager

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1674,101 +1922,294 @@ def _check_sorted_dependencies(self):
 
-            print ''
+        linter_messages = self._lint_all_files()
+        common_messages = LintChecksManager.perform_all_lint_checks(self)
```
Use `super` here

cuichenli replied:
Done

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -2656,15 +2777,7 @@ def perform_all_lint_checks(self):
-            self._check_sorted_dependencies())
-        controller_dependency_messages = (
-            self._match_line_breaks_in_controller_dependencies())
+        common_messages = LintChecksManager.perform_all_lint_checks(self)
```
Ditto

cuichenli replied:
Done

kevinlee12 reviewed (changes requested):
@cuichenli, this is looking really close! two more comments :)

cuichenli committed:
Merge branch 'develop' of github.com:oppia/oppia into 7305-separate-lintchecksmanager

cuichenli committed:
Use super

kevinlee12 reviewed (approved):
lgtm, thanks @cuichenli!

@ankita240796, can you do a final pass at this? thanks!

ankita240796 reviewed (approved):
LGTM, Thanks @cuichenli!
